### ISSUE TITLE
Incorrect "unreachable pattern" warning with opaque constant

### ISSUE THREAD

**Issue by Nadrieril** [rust-lang/rust#78057]
I tried this code ([playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=81b06fa562674b849e41115b8309e43f)):

```rust
#[derive(PartialEq)]
struct Opaque(i32);

impl Eq for Opaque {}

const FOO: Opaque = Opaque(42);

fn foo() {
    match Opaque(0) {
        FOO => {},
        _ => {}
    }
}
```

This emits an "unreachable pattern" warning on the second branch of the match, even though it is clearly reachable. This gives the same warning for current nightly and beta, as well as stable 1.47.0 and 1.40.0.

This was uncovered while discussing https://github.com/rust-lang/rust/pull/77390.

### Meta

`rustc --version --verbose`:
```
rustc 1.47.0 (18bf6b4f0 2020-10-07)
binary: rustc
commit-hash: 18bf6b4f01a6feaf7259ba7cdae58031af1b7b39
commit-date: 2020-10-07
host: x86_64-unknown-linux-gnu
release: 1.47.0
LLVM version: 11.0
```

**Comment by bugadani** [comment]
How is the second branch reachable?

**Comment by Nadrieril** [comment]
A value like `Opaque(0)` would reach it. I do realize that using `FOO` as the match scrutinee is ocnfusing, I'll change that

**Comment by bugadani** [comment]
And now the warning isn't raised ;)

**Comment by bugadani** [comment]
With this example, I get an error: `error: to use a constant of type `Opaque` in a pattern, `Opaque` must be annotated with `#[derive(PartialEq, Eq)]`

I believe the warning is raised because the compiler then assumes you use FOO as a catch-all name instead?

**Comment by Nadrieril** [comment]
Oh, that's an interesting hypothesis. That would mean some part of the code sees this as a const pattern (hence the error) and some other part as a variable binding (hence the warning). I would find that quite surprising honestly.

**Comment by Nadrieril** [comment]
Nah, the bit that emits the error is in `rustc_mir_build/src/thir/pattern/const_to_pat.rs`, and the one that emits the warning is `rustc_mir_build/src/thir/pattern/check_match.rs`. I'm mostly sure that those two take the same inputs and in particular would not reclassify the const as a variable binding

**Comment by bugadani** [comment]
But if I derive `Eq` instead of implementing it like you, it's all good: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c3d83bafde050c32204321a21cf44689

**Comment by Nadrieril** [comment]
Yeah, something happens when the const is considered opaque which triggers this behavior. I'm not sure what exactly, @oli-obk do you know? Maybe returning `Err` after emitting the error in const-to-pat somehow transforms the pattern into a wildcard?

**Comment by Nadrieril** [comment]
Oh wait that's exactly what happens :D 
https://github.com/rust-lang/rust/blob/6af9846fcc8797bf97e9fb387385208c2219f3d0/compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs#L337-L348

@oli-obk I believe to solve that particular issue you'd need to return something else than `PatKind::Wild` here

**Comment by Nadrieril** [comment]
I gave it more thought, and the correct approach seems instead to not even start exhaustiveness checking if `const_to_pat` emits an unrecoverable error.

**Comment by oli-obk** [comment]
How about we introduce a `PatKind::Error` to make these hacky `PatKind::Wild` situations obvious? Then exhaustiveness can just bail out and assume an error has already been reported.

**Comment by Nadrieril** [comment]
There's already code in check_match to track errors and skip exhaustiveness, we could reuse that: https://github.com/rust-lang/rust/blob/2b76c483282f5eae595766fc79ec20818650017e/compiler/rustc_mir_build/src/thir/pattern/check_match.rs#L182
I'll add that to my queue of things to do

**Comment by Nadrieril** [comment]
yeah ok there's a complicated path from `const_to_pat` to `check_match`. `PatKind::Error` sounds much more feasible

**Comment by Taha-Firoz** [comment]
I've hit the same problem but matching against const Uuid's. Is there a work around other than adding an if statement in each arm.

**Comment by Dylan-DPC** [comment]
The code no longer emits this warning for this case.

**Comment by Nadrieril** [comment]
It does, I just tested with the same [playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=81b06fa562674b849e41115b8309e43f). But it also emits an error so maybe the warning doesn't matter too much.
